### DISCUSSION TITLE
Supporting Non-IP Protocols in the Connection object

### DISCUSSION THREAD

**Discussion by timwoj** [zeek/zeek#1383] [Projects]
As part of our continuing efforts to support more protocols, we’re now taking a look at the Connection object and log. There is currently a fixed dependency between that type and the need to have IP-based traffic. Users have been requesting that we support non-IP protocols for many years, and removing this dependency is one step in that work. There are a number of major issues to be resolved here, all of them semi-related to each other:

- How to rework the Conn object to support run-time configurations that aren’t IP-based. Should the existing 5-tuple always exist?
- If new fields are added and the existing 5-tuple is ignored, how should the session code use those fields as the ConnIDKey when comparing connections? Should all of the fields - new and old - be used as a key together?
- How should a user be able to add custom additional fields to both the Conn and ConnIDKey? This should likely be in script land since it allows for run-time and plugin-based changes, but what does that configuration look like?
- How does this affect logging of Conn records? We obviously want to be very careful not to break the existing logging, since it would break a large number of third-party tools.

I’m opening to this discussion to consider the above questions with the community, and see if we can land on something that works for everyone.

One current idea to get the conversation started is to change the ConnIDKey to just be a Record type (and a RecordVal in the C++ code). The record type would be defined by default to contain the existing 5-tuple. The Connection record type would contain one of these records. This would be modifiable from script-land, and we can use the RecordVal underneath for the key since it’s easily hashed.

Related github issue: https://github.com/zeek/zeek/issues/284

**Comment by JustinAzoff** [comment]
I see you mention additional fields a bunch, but there's also the case of fewer fields.  ipsec and icmp don't really have ports and are really only a 3-tuple.  We currently overload the ports for icmp to be the type and code.

**Reply by timwoj** [reply]
In this case the 5-tuple would still exist unused for the most part, but we could easily add extra fields to cover the ICMP type and code. This probably falls under the question about supporting existing tools though too.

**Comment by overmetal61** [comment]
My Twocents..   
I don't personally find the 5 tuple valuable.  My understanding is that it can be used for correlation.   I would rather see the connection data enriched within each particular log and de-duplicated..  And conn.log become a catch-all for whatever doesn't fall into some other log type.

**Comment by J-Gras** [comment]
From what I understand, the idea is to allow new protocols to introduce their notion of a connection by making the key that identifies a connection configurable. I don't think that this approach can be realized in a way that a conn.log will contain TCP connections as well as connections for a new protocol. Rather each session-aware protocol could define "its own conn.log".

I think it might make sense to consider configurable connection IDs and new types of connections separately. I can imagine to introduce the abstract concept of a session that is identified by some key. In case of TCP the key could be the 5-tuple. Adding new protocols, it would be up to the developer of the new protocol analyzer to come up with a session identifier. A session manager would allow to manage sessions per protocol. It would schedule packets to sessions, manage per session state and expiration. How to process packets of the same session, the definition of session state and timeouts would be defined in the scope of the session analyzer implementation (e.g., the TCP analyzer). Then the session analyzer is also responsible for generating a protocol specific session log. In case of TCP this would be todays conn.log.

The next question could be how to correlate things if we there are multiple types of sessions. An idea could be that sessions are hierarchical and may contain other sessions. This would be kind of a generalization of the current concept for tunneling. A TCP session could contain multiple HTTP sessions. This could also help to properly distinguish between different contexts: A packet from 1.2.3.4 to 1.2.3.5 from the inside of some VXLAN tunnel might be in another session scope than packets outside of that tunnel.

Finally, configurable session IDs would just be a matter of the session analyzer. Developers could decide to hard-code how sessions are identified in their session analyzers or read in script-land configuration. For example, in case of TCP the default session identifier would be the 5-tuple but there could be an option to include VLAN tags into the identifier if present. An important question is how to make meta data of lower layers available to higher-layer analyzers (e.g., IPs for the TCP session analyzer's 5-tuple)? This was already discussed in the context of the packet analyzer changes.

**Reply by rsmmr** [reply]
> I don't think that this approach can be realized in a way that a conn.log will contain TCP connections as well as connections for a new protocol. Rather each session-aware protocol could define "its own conn.log".

I really like this observation.  Keeping non-IP stuff out of existing logs simplifies the scriptland-side of this substantially, and it makes sense from a user perspective too: `conn.log` would essentially keep meaning the TCP/IP suite (just as today), and new stuff would show up elsewhere and with its own custom semantics. 

With that, the suggestion of a generic session manager makes sense then too. API-wise, the session identifier could just be HashKey that the protocol analyzers provide, based whatever sessions semantics they want.

Also agree that configurable session IDs then become an orthogonal issue to be solved by the corresponding analyzer.

**Reply by timwoj** [reply]
In that model, would the existing `Connection` object remain IP-specific? I could see there being some sort of base class that it inherits from for storing a generic object in the session manager.

**Comment by duffy-ocraven** [comment]
Considering that #284 is about to celebrate its second birthday, and doing anything about this has not yet gotten off the dime, I suspect the community has some significant hesitation-paralysis. I would suggest doing a first alteration which preserves the 5-tuple (e.g. with ICMP, BACnet, or VLAN), thus minimizing the chance of breaking anything; all consumers will basically get the format they are used to, just the values/numbers are different. That will give us some feedback-from-experience regarding how large and where were the needed changes in the implementation, and whether doing that contributed anything.

**Comment by duffy-ocraven** [comment]
J-Gras above wrote "... generalization of the current concept for tunneling. A TCP session could contain multiple HTTP sessions..." and that reminded me of the formidable prospect of full SCTP support. I think separately from this ticket, which should stay devoted to Conn record evolution, a separate ticket should undertake SCTP support, and move any consideration of "concept for tunneling" over there to see what is learned from that undertaking. It will be valuable to find out where in the process of successively fleshing-out SCTP coverage, a day-hike turns into El Capitan.
